src/registrar/tests/services/test_dns_host_service.py

```diff
@@ -131,3 +146,112 @@ def test_create_record_failure(self, mock_create_dns_record):
         with self.assertRaises(APIError) as context:
             self.service.create_record(zone_id, record_data)
         self.assertIn("Bad request: missing name", str(context.exception))
+
+
+class TestDnsHostServiceDB(TestCase):
+    def setUp(self):
+        self.vendor = DnsVendor.objects.get(name=DnsVendor.CF)
+        mock_client = Mock()
+        self.service = DnsHostService(client=mock_client)
+
+    def tearDown(self):
+        DnsVendor.objects.all().delete()
+        VendorDnsAccount.objects.all().delete()
+        DnsAccount.objects.all().delete()
+        Join.objects.all().delete()
+
+    def test_save_db_account_success(self):
+        # Dummy JSON data from API
+        account_data = {"result": {"id": "12345", "name": "Account for test.gov", "created_on": "2024-01-02T03:04:05Z"}}
+
+        # Validate that the method returns the vendor account ID
+        returned_id = self.service.save_db_account(account_data)
+        self.assertEqual(returned_id, "12345")
+
+        # Validate there's one VendorDnsAccount row with the external id and the CF Vendor
+        self.assertEqual(VendorDnsAccount.objects.count(), 1)
+        vendor_acc = VendorDnsAccount.objects.get(x_account_id="12345")
+        self.assertEqual(vendor_acc.dns_vendor, self.vendor)
+
+        # Validate there's one DnsAccount row with the given name
+        self.assertEqual(DnsAccount.objects.count(), 1)
+        dns_acc = DnsAccount.objects.get(name="Account for test.gov")
+
+        # Testing a join row for DnsAccount_VendorDnsAccount
+        self.assertEqual(Join.objects.count(), 1)
+        join = Join.objects.get()
+        self.assertEqual(join.dns_account, dns_acc)
+        self.assertEqual(join.vendor_dns_account, vendor_acc)
+
+    def test_save_db_account_with_error_fails(self):
+        account_data = {"result": {"id": "FAIL1", "name": "Failed Test Account", "created_on": "2024-01-02T03:04:05Z"}}
+
+        # patch() temporarily replaces VendorDnsAccount.objects.create() with a fake version that raises
+        # an integrity error mid-transcation
+        with patch("registrar.models.VendorDnsAccount.objects.create", side_effect=IntegrityError("simulated failure")):
+            with self.assertRaises(IntegrityError):
+                self.service.save_db_account(account_data)
+
+        # Ensure that no database rows are created across our tables (since the transaction failed).
```

Comment:
SImilarly here, rather than checking the total count in tables, I'd check that the count did not change. So I'd check the total count before calling `save_db_account`, and then assert against the total count after calling `save_db_account`